include/swift/Reflection/ReflectionContext.h

```diff
@@ -415,24 +424,28 @@ class ReflectionContext
 
     assert(SecIdx < SecHdrVec.size() && "malformed ELF object");
 
-    const typename T::Section *SecHdrStrTab = SecHdrVec[SecIdx];
+    std::unique_ptr<const typename T::Section> &SecHdrStrTab =
+        SecHdrVec[SecIdx];
     typename T::Offset StrTabOffset = SecHdrStrTab->sh_offset;
     typename T::Size StrTabSize = SecHdrStrTab->sh_size;
 
-    auto StrTabStart =
-        RemoteAddress(ImageStart.getAddressData() + StrTabOffset);
-    auto StrTabBuf = this->getReader().readBytes(StrTabStart, StrTabSize);
-    auto StrTab = reinterpret_cast<const char *>(StrTabBuf.get());
+    // The string table is also not mapped in the child process,
+    // read its contents from the file buffer.
+    uint64_t StrTabStart = FileBufferStartAddress + StrTabOffset;
+    std::vector<char> StrTab(StrTabSize);
+    memcpy(StrTab.data(), (void *)(StrTabStart), StrTabSize);
 
     auto findELFSectionByName = [&](std::string Name)
         -> std::pair<RemoteRef<void>, uint64_t> {
       // Now for all the sections, find their name.
-      for (const typename T::Section *Hdr : SecHdrVec) {
+      for (auto &&Hdr : SecHdrVec) {
         uint32_t Offset = Hdr->sh_name;
-        auto SecName = std::string(StrTab + Offset);
+        char *SecName = StrTab.data() + Offset;
```

Comment:
Since we are going to compare it against a std::string anyway, we might as well keep the type as a std::string here.
Here is another opportunity to crash on malformed input. What do you think about using a std::string constructor that limits the size of the string to (StrTab+length(StrTab)-Offset)?

Reply:
(malformed=missing NUL-terminator)

Reply:
It looks like the string constructor that takes a `char *` and a size ignores the null terminator, setting the string to the size passed in. What if we build a string using just the `char *`, then check its size is within limits, and skip (or assert) if it isn't.

Reply:
I'm worried that the constructor would run past the end of the section while scanning for the NUL. That's ugly, but we could either manually scan the string for a NUL first or strncpy it into a separate buffer.

Reply:
Ok, I've updated the code to check the size is valid beforehand.

Reply:
There's a libc function called `strnlen_s( const char *str, size_t strsz)` that we can call for this.